Reset-Office365Tenant/Office365.psm1

```diff
@@ -0,0 +1,377 @@
+        return $hasImpersonationRights
+    }
+    catch
+    {
+        DisplayCompletedOperation "FAIL" Red
+        Write-Host -ForegroundColor Red "ERROR: Get-ManagementRoleAssignment failed with $_"
+        return $null
+    }
+}
+
+############################################################
+# Enable OrganizationCustomization for the tenant if not hydrated
+############################################################
+function Enable-TenantCustomization()
+{
+    try
+    {
+        DisplayInProgressOperation "Enable OrganizationCustomization"
+        Enable-OrganizationCustomization -ErrorAction silentlycontinue
+        DisplayCompletedOperation "OK" Green
+        return $true
+    }
+    catch
+    {
+        DisplayCompletedOperation "FAIL" Red
+        Write-Host -ForegroundColor Red "ERROR: Enable-OrganizationCustomization failed with $_"
+        return $false
+    }
+}
+
+############################################################
+# Add Impersonation Rights to a user
+############################################################
+function Add-ImpersonationRights($username)
+{
+    try
+    {
+        DisplayInProgressOperation "Set Impersonation rights"
+        New-ManagementRoleAssignment -Role "ApplicationImpersonation" -User  $username
+        DisplayCompletedOperation "OK" Green
+        return $true
+    }
+    catch
+    {
+        DisplayCompletedOperation "FAIL" Red
+        Write-Host -ForegroundColor Red "ERROR: Set-ManagementRoleAssignment failed with $_"
+        return $false
+    }
+}
+
+################################################################################
+# Confirms if user has Impersonation Rights
+################################################################################
+function Confirm-ImpersonationRights($username)
+{
+    # Test if tenant is dehydrated
+    $isDehydrated = Test-IsTenantDehydrated
+
+    # Fail if we cannot get this information
+    if ($null -eq $isDehydrated)
+    {
+        return $false
+    }
+    
+    # Enable-OrganizationCustomization if tenant is dehydrated
+    if ($true -eq $isDehydrated)
+    {
+        if ($false -eq $(Enable-TenantCustomization))
+        {
+            return $false
+        }
+    }
+
+    # Test if user has Impersonation rights
+    $hasRights = Test-HasImpersonationRights -username $username
+    if ($false -eq $hasRights)
+    {
+        # Add Impersonation rights
+        return Add-ImpersonationRights -username $username
+    }
+    return $true
+}
+
+################################################################################
+# Reset Office365 Tenant
+################################################################################
+function Reset-Office365Tenant($credentials, $defaultPassword, $maxUser)
+{
+    if ($null -eq $credentials)
+    {
+        Write-Host "Ask for credentials"
+        $credentials = Get-Credentials -Message "Please insert your Office365 Admin Credentials" 
```

Comment:
Tow things here: 1. "Get-Credentials" is not the name of a commandlet, nor do you define it anywhere, so this operation will fail. (Are you looking for "Get-Credential"? 2. Once that is fixed, what if the user clicks the "X" or enters the wrong credentials? I advise repeating the credential request, probably with a loop, until the credentials are not null. It may also be beneficial to add a check of some sort to see if they are correct.

Reply:
Fixed the typo. I also wrapped this into another function. Although, I actually want that if a user press "esc" the windows goes away and the script stops.